<a href="https://colab.research.google.com/github/wbwatkinson/udemy_numpy_stack/blob/main/numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Section 1 of:

https://www.udemy.com/deep-learning-prerequisites-the-numpy-stack-in-python/

https://deeplearningcourses.com/c/deep-learning-prerequisites-the-numpy-stack-in-python

## Arrays vs Lists

In [21]:
import numpy as np

In [22]:
L = [1,2,3]

In [23]:
A = np.array([1,2,3])

In [24]:
for e in L:
  print(e)

1
2
3


In [25]:
for e in A:
  print(e)

1
2
3


In [26]:
L.append(4)

In [27]:
L

[1, 2, 3, 4]

In [28]:
# A.append(4) # doesn't work

In [29]:
L + [5]

[1, 2, 3, 4, 5]

In [30]:
A + np.array([4])

array([5, 6, 7])

In [31]:
A + np.array([4,5,6])

array([5, 7, 9])

In [32]:
# A + np.array([4,5]) # doesn't work

In [33]:
2 * A

array([2, 4, 6])

In [34]:
2 * L

[1, 2, 3, 4, 1, 2, 3, 4]

In [35]:
L + L

[1, 2, 3, 4, 1, 2, 3, 4]

In [36]:
L2 = []
for e in L:
  L2.append(e + 3)

In [37]:
L2

[4, 5, 6, 7]

In [38]:
L2 = [e + 3 for e in L]

In [39]:
L2

[4, 5, 6, 7]

In [40]:
L2 = []
for e in L:
  L2.append(e**2)

In [41]:
L2

[1, 4, 9, 16]

In [42]:
A**2

array([1, 4, 9])

In [43]:
np.sqrt(A)

array([1.        , 1.41421356, 1.73205081])

In [44]:
np.log(A)

array([0.        , 0.69314718, 1.09861229])

In [45]:
np.exp(A)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [46]:
np.tanh(A)

array([0.76159416, 0.96402758, 0.99505475])

## The Dot Product

$$ a \cdot b = a^T b = \sum_{d=1}^D a_d b_d $$

In [47]:
a = np.array([1,2])
b = np.array([3,4])

In [48]:
dot = 0
for e, f in zip(a, b):
  dot += e*f
dot

11

In [49]:
# use integer index
dot = 0
for i in range(len(a)):
  dot += a[i] * b[i]
dot

11

In [50]:
a * b

array([3, 8])

In [51]:
np.sum(a * b)

11

In [52]:
(a * b).sum()

11

In [53]:
np.dot(a, b)

11

In [54]:
a.dot(b)

11

In [55]:
b.dot(a)

11

In [56]:
a @ b

11

$$ a^T b = \left\lVert a \right\rVert \left\lVert b \right\rVert \cos \theta_{ab}$$

$$ \cos \theta_{ab} = \frac{a^T b}{\left\lVert a \right\rVert \left\lVert b \right\rVert} $$

$$ \left\lVert a \right\rVert = \sqrt{\sum_{d=1}^D a_d^2 } $$

In [57]:
amag = np.sqrt((a * a).sum())
amag

2.23606797749979

In [58]:
amag = np.linalg.norm(amag)
amag

2.23606797749979

In [59]:
cosangle = a.dot(b) / (np.linalg.norm(a) * np.linalg.norm(b))

In [60]:
cosangle

0.9838699100999074

In [61]:
angle = np.arccos(cosangle)
angle

0.17985349979247847

## Speed Test

In [62]:
## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.random.randn(100)
b = np.random.randn(100)
T = 100000

def slow_dot_product(a, b):
  result = 0
  for e, f in zip(a, b):
    result += e*f
  return result

t0 = datetime.now()
for t in range(T):
  slow_dot_product(a, b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
  a.dot(b)
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())

dt1 / dt2: 60.949655582957725


## Matrices

In [63]:
L = [[1,2], [3,4]]
L

[[1, 2], [3, 4]]

In [64]:
L[0]

[1, 2]

In [65]:
L[0][1]

2

In [67]:
A = np.array([[1,2],[3,4]])
A

array([[1, 2],
       [3, 4]])

In [68]:
A[0][1]

2

In [69]:
A[0,1]

2

In [70]:
A[:,0] # returns the column at index 0 (select every row of column 0)

array([1, 3])

In [71]:
A.T

array([[1, 3],
       [2, 4]])

In [72]:
np.exp(A)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [73]:
np.exp(L) # pass a list into numpy operation

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [74]:
B = np.array([[1,2,3],[4,5,6]])
B

array([[1, 2, 3],
       [4, 5, 6]])

In [75]:
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [77]:
# A.dot(B.T) # doesn't work, not aligned

In [78]:
np.linalg.det(A)

-2.0000000000000004

In [79]:
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [80]:
np.linalg.inv(A).dot(A)

array([[1.00000000e+00, 0.00000000e+00],
       [1.11022302e-16, 1.00000000e+00]])

In [81]:
np.trace(A)

5

In [82]:
np.diag(A) # get the diagonal elements of A

array([1, 4])

In [83]:
np.diag([1,4]) # create a diagonal matrix with 1 and 4 on the diagonal

array([[1, 0],
       [0, 4]])

In [84]:
np.linalg.eig(A) # two arrays: eigen values, eigen vectors in a matrix

(array([-0.37228132,  5.37228132]), array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

In [94]:
Lam, V = np.linalg.eig(A)

array([[-0.37228132,  0.        ],
       [ 0.        ,  5.37228132]])

In [86]:
V[:,0] * Lam[0] == A @ V[:,0] # eigen vector times eigen value should equal matrix times eigen vector

array([ True, False])

In [88]:
V[:,0] * Lam[0], A @ V[:,0]

(array([ 0.30697009, -0.21062466]), array([ 0.30697009, -0.21062466]))

In [89]:
np.allclose(V[:,0] * Lam[0], A @ V[:,0])

True

In [95]:
np.allclose(V @ np.diag(Lam), A @ V)

True

# Solving Linear Systems

The admission fee at a small fair is \\$1.50 for children and \\$4.00 for adults. On a certain day, 2200 people enter the fair, and \\$5050 is collected. How many children and how many adults attended?

\begin{align}
  x_1 + x_2 & = 2200 \\
  1.5x_1 + 4x_2 & = 5500
\end{align}

$$ x = \left ( \begin{matrix} x_1 \\ x_2 \end{matrix} \right ) , 
A = \left ( \begin{matrix} 1 & 1 \\ 1.5 & 4 \end{matrix} \right ) ,
b = \left ( \begin{matrix} 2200 \\ 5050 \end{matrix} \right ) $$

$$ Ax = b \Leftrightarrow x = A^{-1}b $$

In [100]:
A = np.array([[1,1], [1.5, 4]])
b = np.array([2200, 5050])

In [101]:
np.linalg.solve(A, b) # faster
np.linalg.inv(A).dot(b) # don't use this, it's slower

array([1500.,  700.])

# Generating Data

In [102]:
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [103]:
np.ones((2,3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [104]:
10 * np.ones((2,3))

array([[10., 10., 10.],
       [10., 10., 10.]])

In [105]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [106]:
np.random.random()

0.43054462933397575

In [107]:
np.random.random((2, 3)) # uniform [0,1] distribution

array([[0.0059372 , 0.29035161, 0.31157017],
       [0.71934622, 0.17274325, 0.41302513]])

In [110]:
np.random.randn(2,3) # gaussian (normal) distribution mean = 0, variance = 1

array([[ 0.50084273,  1.31011758, -1.0498433 ],
       [ 1.73265316, -1.37305146, -0.54179276]])

In [112]:
R = np.random.randn(10000)

In [113]:
R.mean()

0.013716220584741427

In [114]:
np.mean(R)

0.013716220584741427

In [115]:
R.var()

0.9947986995925902

In [116]:
R.std() # standard deviation is the square root of the variance

0.9973959592822653

In [120]:
R = np.random.randn(10000,3)

In [121]:
R.mean(axis = 0) # the mean of each column

array([-0.01643795, -0.00137608, -0.00059358])

In [122]:
R.mean(axis = 1) # the mean of each row

array([ 1.02765338,  0.00976358, -0.25749441, ...,  0.04328973,
       -0.37537898,  0.83809182])

In [123]:
R.mean(axis=1).shape

(10000,)

In [ ]:
np.cov(R) # treats each column as a vector observation

In [126]:
np.cov(R).shape # analog of variance with vectors is covariance

(10000, 10000)

In [127]:
np.cov(R.T) # take transpose first to treat each row as an observation

array([[ 0.98076511,  0.00407853,  0.00375116],
       [ 0.00407853,  1.01678517, -0.00657914],
       [ 0.00375116, -0.00657914,  1.00419473]])

In [128]:
np.cov(R, rowvar = False) # passing rowvar = False has same effect as transposing R

array([[ 0.98076511,  0.00407853,  0.00375116],
       [ 0.00407853,  1.01678517, -0.00657914],
       [ 0.00375116, -0.00657914,  1.00419473]])

In [129]:
np.random.randint(0,10,size=(3,3))

array([[6, 7, 4],
       [9, 4, 1],
       [2, 5, 7]])

In [130]:
np.random.choice(10, size=(3,3))

array([[4, 6, 6],
       [9, 0, 5],
       [9, 5, 2]])

# Numpy Exercise

Do a speed test for matrix multiplication

In [149]:
## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.random.randint(0,10, size=(3,3))
b = np.random.randint(0,10, size=(3,3))
T = 100000

def list_comp_matrix_multiplication(a, b):
  return [[sum([a*b for (a,b) in zip(row, col)]) for col in zip(*b)] for row in a]

def for_loop_matrix_multiply(a, b):
  result = np.zeros((a.shape[0], b.shape[1]))

  for i, row in enumerate(a):
    for j, col in enumerate(b.T):
      # result[i,j] = slow_dot_product(row, col) #row @ col
      result[i, j] = sum([row*col for (row, col) in zip(row,col)])

  return result


def slow_dot_product(a, b):
  result = 0
  for e, f in zip(a, b):
    result += e*f
  return result

t0 = datetime.now()
for t in range(T):
  for_loop_matrix_multiply(a,b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
  a @ b
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())
a, b, a.dot(b), a @ b, for_loop_matrix_multiply(a, b)

dt1 / dt2: 28.682467101818563


(array([[2, 1, 5],
        [7, 8, 0],
        [4, 9, 9]]), array([[8, 2, 0],
        [9, 2, 2],
        [7, 5, 5]]), array([[ 60,  31,  27],
        [128,  30,  16],
        [176,  71,  63]]), array([[ 60,  31,  27],
        [128,  30,  16],
        [176,  71,  63]]), array([[ 60.,  31.,  27.],
        [128.,  30.,  16.],
        [176.,  71.,  63.]]))